text/0000-localizing-api-messages.md

```diff
@@ -0,0 +1,135 @@
+only clients that are built using Stripes. As a result, we are recommending that we go with Option 4._**
+</span>.
+
+### Option 1
+Distribute them over the existing front-end modules:
+
+ui-checkin with translation key ui-checkin.mod-circulation.itemNotFound
+ui-requests with translation key ui-requests.mod-circulation.patronHasItemOnLoad
+
+*Pros*
+- Uses existing front-end modules
+
+*Cons*
+- Cannot disable a front-end module for a tenant if it hosts a mod-circulation translation key that is needed by another enabled front-end module.
+- Not all originating messages can be mapped to a frontend module 
+
+### Option 2a
+For each back-end module with translation keys create a dedicated front-end module with language files:
+
+For example create ui-mod-circulation module, and use translation keys
+
+ui-mod-circulation.itemNotFound
+ui-mod-circulation.patronHasItemOnLoad
+
+*Pros*
+- Any front-end module that uses mod-circulation can require ui-mod-circulation to load the translation files.
+- All mod-circulation translation strings go into a single module.
+*Cons*
+- Additional overhead in managing more ui modules
+- Not all originating messages can be mapped to a frontend module
+
+### Option 2b
+For each back-end interface with translation keys create a dedicated front-end module with language files:
+
+For example create i18n-circulation module, and use translation keys
+
+i18n-circulation.itemNotFound
+i18n-circulation.patronHasItemOnLoad
+
+*Pros*
+- Any front-end module that uses the circulation API can require i18n-circulation to load the translation files.
+- All circulation translation strings go into a single module.
+*Cons*
+- Additional overhead in managing more ui modules
+- Not all originating messages can be mapped to a frontend module
+
+### Option 3
+The back-end module hosts the language files.
+
+mod-circulation creates a translations/mod-circulation/ directory.
+
+Stripes fetches the translation files from the back-end module and can process them the same way as translation files from frone-end modules.
+
+*Pros*
+- Software and language files are in the same repository. Options 1 and 2 require to add the message string to some other repository using a second pull request.
+
+*Cons*
+- Is this technically feasible? How can Stripes fetch languages files from a non-Stripes module?
+
+### Option 4:
+When calling the back-end pass the required locale. The back-end maintains the translations files, replaces the placeholders and puts the final string into the "message" property. A lang query parameter that many FOLIO API interfaces already have or an "accept-language" header line might be used.
+
+*Pros*
+- Java code and translation files live in the same repository. No work for the front-end.
+- This is a well known pattern in the Java community
+
+*Cons*
+- None
+
+## Risks and Drawbacks
+
+None that we can think of. The pattern we are adopting is a standard pattern in the java community
```

Comment:
Similarly to my comment about cons, it is unlikely that there are no drawbacks to an approach at all.

Is this the section to talk about the impact of not considering the process for changing the translations?

To me, that area presents the most potential for trade offs and drawbacks that don't seem to be mentioned in this document yet.

Reply:
One drawback is that the translation proposal is specific to one programming language. If we have a long-term vision of multiple languages, that might be an issue and new conventions would need to be implemented. Another drawback I infer is that translators have to hunt down every module.

One option that has not been considered is a central service to store localization strings. That would be available to modules written in whatever programming language. As a side-effect, it would permit reusing message text across modules.

Conceivably, back-end _or_ front-end modules could call it with a key and a language and get back the needed string. A module could cache translations as they are called, and the service could issue an alert when there are new translations and the cache needs to be refreshed.

Translators have one central storage to see everything that needs to be translated. If a key were constructed via a naming convention, that could intentionally include info about the use context to translators.

Reply:
Our [module submission guidelines state that "Back-end modules are based on Maven/JDK 11 and provide a Dockerfile"](https://github.com/folio-org/tech-council/blame/e72b82ba80aa51a5e8c5e989d2bafb86d7488b98/MODULE_EVALUATION_TEMPLATE.MD#L31) thus I see no problem with JDK-specific guidance here.

Reply:
> Our [module submission guidelines state that "Back-end modules are based on Maven/JDK 11 and provide a Dockerfile"](https://github.com/folio-org/tech-council/blame/e72b82ba80aa51a5e8c5e989d2bafb86d7488b98/MODULE_EVALUATION_TEMPLATE.MD#L31) thus I see no problem with JDK-specific guidance here.

I think this might be ones of those cases where we accepted some limitations in the criteria due to it only applying to  new modules and the time constraints we had to put this together.

Assuming this decision will apply to all existing modules, it will need to cater to (at least) the following languages / runtimes / build systems:
* java / maven
* node.js / yarn (mod-graphql)
* grails / gradle (mod-agreements, mod-licences etc)
* perl / make (net-Z39-50-FOLIO)

At the very least, I think it could be good to distinguish (e.g. separate sections / headings) the general standard from the language / runtime specific guidance.